### ISSUE TITLE
Question about Architecture

### ISSUE THREAD

**Issue by ghost** [vimeo/PlayerKit#69]
Great library! I have a question about architecture. I checked out your included sample project and this is how the view controller is setup:
```
class PlayerViewController: UIViewController, PlayerDelegate {
    private let player = RegularPlayer()
}
```
Notice how the view controller does not use dependency injection. The view controller simply creates its player directly.

My question, is let's say I create my own custom players using `PlayerKit` such as:
```
final class MusicPlayer: NSObject, Player, ProvidesView { }
final class AudioBookPlayer: NSObject, Player, ProvidesView { }
```

Would you suggest having custom view controllers for each player such as the following?
```
final class MusicPlayerViewController: UIViewController, PlayerDelegate {
    private let player = MusicPlayer()
}

final class AudioBookPlayerViewController: UIViewController, PlayerDelegate {
    private let player = AudioBookPlayer()
}
```

Basically, each view controller creates it's player directly.

Does this make sense? Any thoughts?

**Comment by rpassis** [comment]
Hey @johnswifttt,

Please don't take the example too literally, that's just a simple exercise in order to get you going and using the player quickly. For a more testable, reusable and extendable solution I'd definitely recommend you inject the player into the view controller.

How you do that will generally depend on the general architecture of your app, how your controllers are instantiated and where from but as a general rule I personally prefer having static factory methods that take dependencies as parameters and return a configured view controller to the calling site. 

Here's one possible implementation following your example above and using the same view controller for different player types.

```swift
protocol Player { // Omitted for brevity }
protocol ProvidesView { // Omitted for brevity }

class MusicPlayer: Player, ProvidesView { // Omitted for brevity }
class AudioBookPlayer: Player, ProvidesView { // Omitted for brevity }

class MyViewController<T: Player & ProvidesView>: UIViewController {

    var player: T!
    
    static func configured(with player: T) -> MyViewController {
        let vc = MyViewController()
        vc.player = player
        return vc
    }
}

let musicPlayer = MusicPlayer()
let musicVc = MyViewController.configured(with: musicPlayer)

let audioBookPlayer = AudioBookPlayer()
let audioBookPlayerVc = MyViewController.configured(with: audioBookPlayer)
```

I hope that helps!

**Comment by ghost** [comment]
@rpassis  Oh I see, that's a very clean approach! Thank you for such a quick reply and sample code!

I have one more question, if you don't mind 🙂 (it's kind of related to this library)

Right now, our app only has `Video`’s. We are expanding to `Song`’s and soon other media such as `Podcast`’s, etc. Also, right now, there is a `PlayerViewController`. This is coupled with `Video` objects tightly. The configure property for this view controller is like so:

```
func inject(videos: [Video], currentVideoIndex: Int, videoPlayer: VideoPlaybackManager) { … }
```

`PlayerViewController` has playback controls such as a play button, pause button, next/previous button, and others. We want to use this **same layout** of controls for `Song`’s and other media. There is a XIB for this: `PlaybackControlsView`. This is the only thing in common of how we would display these objects to the user. Basically, for Videos we would show videos, but as for Songs/Podcasts we would instead show an image.

I'm not sure if I should have only 1 view controller `PlayerViewController` that plays all types of media. So I would use dependency injection to inject:
1. The player (like you showed above)
2. The API strategy (what API to use to fetch either Video, Songs, or Podcasts)
3. View strategy (what view to display: a video or image).. depends on if its is a Video or Song/Podcast

Or should I have a `VideoPlayerViewController` and `AudioPlayerViewController` instead of just 1 `PlayerViewController`?

**Comment by rpassis** [comment]
I can't answer those questions for you without knowing a lot more about your project and codebase but if you are looking for an *opinion* based on what you described above, it feels like a single view controller should be able to handle all of those asset types without having any knowledge of what their concrete types are.

**Comment by ghost** [comment]
@rpassis Got it, thanks for your opinion, that helps. Also, thanks for the quick reply! And great library, it helps a lot in terms of architecture
